projects/storefrontlib/src/cms-components/cart/cart-coupon/cart-coupon.component.html

```diff
@@ -0,0 +1,43 @@
+<ng-container *ngIf="cart$ | async as cart">
+  <div class="cx-cart-coupon-title" data-test="title-coupon">

```

Comment:
We should not use attributes that facilitate address test cases only

Reply:
> Hi Stan, may I know the reason? due to we are using BDD mode, and want to decouple the implementation of component. because we don't know which css will be used in component when writing UT

I discussed this with Stan, and this is for sure a coding rule regarding tests they agreed on, but it's not mentioned in coding guidelines. Stan will go to enhance the guideline.

Reply:
Hi Kate, I think the main reason is, that we don't have reliable way (yet) to strip them on production, which technically means we would be slowing down our live app (also making it bigger) for the sake of little bit easier testability. Another one is, that this additional markup which exists for the sole purpose of testing, add another level of "maintenance" outside of test cases itself. 

I think unit tests are very low level, so you need to have quite detailed component specs to test desired functionality. Will it be a general sketch (where you know, that you have sth like 'title-coupon') or actual template (`html` + `css`) it depends, but in the end, I don't think you are violating any idea of TDD/DBB if you test correct behaviour against specified component UI (markup with styles). 

When styles are part of the component's public specification (as in our case), it's even more valid, because customers can target those css classes and changing them would mean a breaking change, so IMO unit tests should guard us and fail when someone will change the css class of some element (which won't be a case when using test selectors).

Reply:
Hi Stan, thanks for your explanation, I know adding an additional attribute for test only has pros and cons. I really understand what you are worry about. I am writing an email to ask for our quality architect stephen palmer's suggestion. and will take a look at babel-plugin-react-remove-properties to see whether can compile those attributes away.

Reply:
I don't think that plugin will work (Angular doesn't use babel nor jsx). Also, please take into account that the compilation results here are libraries, so to facilitate those selectors also in e2e tests (which should be possible if we are using them in the first place, just for consistency) makes it even more complicated: you would have to compile two types of libraries, one with selectors and one without them, to use them properly, which raises another set of pros and cons (testing against not exactly the same as final libraries, etc.).

Reply:
Fixed